ansible_base/resource_registry/registry.py

```diff
@@ -0,0 +1,131 @@
+SharedResource = namedtuple("SharedResource", ["serializer", "is_provider"])
+
+
+def get_concrete_model(model):
+    _model = model
+    while _model._meta.proxy_for_model:
+        _model = _model._meta.proxy_for_model
+
+    return _model
+
+
+class ServiceAPIConfig:
+    """
+    This will be the interface for configuring the resource registry for each service.
+    """
+
+    service_type = None
+
+
+class ResourceInspector:
+    def __init__(self, urlpatterns=None):
+        from rest_framework.schemas.generators import BaseSchemaGenerator
+
+        schema_generator = BaseSchemaGenerator()
+
+        schema_generator._initialise_endpoints()
+        self.model_map = {}
+
+        for path, method, callback in schema_generator.endpoints:
+            view = schema_generator.create_view(callback, method)
+
+            if hasattr(view, "get_serializer_class"):
+                try:
+                    serializer_class = view.get_serializer_class()
+                    if hasattr(serializer_class, "Meta"):
+                        model = get_concrete_model(serializer_class.Meta.model)
+                        label = model._meta.label
+                        if label not in self.model_map:
+                            self.model_map[label] = {}
+
+                        if hasattr(view, "action"):
+                            action = view.action
+                            if action == "partial_update":
+                                action = "update"
+                            if action not in self.model_map[label]:
+                                self.model_map[label][action] = []
+                            self.model_map[label][action].append((method, path))
+                except:  # noqa E722
+                    pass
+
+
+class ResourceConfig:
+    model_label = None
+    model = None
+    externally_managed = None
+    managed_serializer = None
+    parent_resources = None
+    actions = None
+    name_field = None
+
+    def __init__(self, model, shared_resource: SharedResource = None, parent_resources: List[ParentResource] = None, name_field: str = None):
+        model = get_concrete_model(model)
+        self.model_label = model._meta.label
+
+        managed_serializer = None
+        externally_managed = False
+        if name_field is None:
+            name_field = "name"
+
+        if shared_resource:
+            managed_serializer = shared_resource.serializer
+            externally_managed = not shared_resource.is_provider
+
+        parent_map = {}
+        if parent_resources:
+            for parent in parent_resources:
+                parent_map[parent.model._meta.label] = parent
+
+        self.model = model
+        self.externally_managed = externally_managed
+        self.managed_serializer = managed_serializer
+        self.parent_resources = parent_map
+        self.actions = ResourceInspector().model_map.get(self.model_label, {})
+        self.name_field = name_field
```

Comment:
Should we validate that the model actually has this name_field? We don't catch an exception or default the name when we try to get the name_field up above where we get the name_field from the model:
```
resource.name = getattr(obj, resource_config.name_field)
```

Reply:
I refactored this so that the update code is cleaner, but if the attribute doesn't exist, we'll just set the name to None. I like defaulting this to "name" since that's the most common field name, but not all resources will necessarily have a name.